src/patterns.md

```diff
@@ -0,0 +1,646 @@
+> &nbsp;&nbsp; | [_PathPattern_]
+
+Patterns in Rust are used to match values against structures and to,
+optionally, bind variables to values inside these structures. They are also
+used in variable declarations and function/closure parameters, though in these
+cases most of the time they are simply used as an identifier that binds to a
+value.
+
+For example, the pattern used in:
+
+```rust
+# struct Car;
+# struct Computer;
+# struct Person {
+#     name: String,
+#     car: Option<Car>,
+#     computer: Option<Computer>,
+#     age: u8,
+# }
+# let person = Person {
+#     name: String::from("John"),
+#     car: Some(Car),
+#     computer: None,
+#     age: 15,
+# };
+if let
+    Person {
+        car: Some(_),
+        age: person_age @ 13..=19,
+        name: ref person_name,
+        ..
+    } = person
+{
+    println!("{} has a car and is {} years old.", person_name, person_age);
+}
+```
+does four things:
+
+* Tests if `person` has the `car` field filled with something.
+* Tests if the person's `age` field is between 13 and 19, and binds its value to
+  the `person_age` variable.
+* Binds a reference to the `name` field to the variable `person_name`.
+* Ignores the rest of the fields of `person`, i.e., they can have any value and
+  are not bound to any variables.
+
+Patterns are used in:
+
+* [`let` declarations](statements.html#let-statements)
+* [Function](items/functions.html) and [closure](expressions/closure-expr.html)
+  parameters
+* [`match` expressions](expressions/match-expr.html)
+* [`if let` expressions](expressions/if-expr.html)
+* [`while let` expressions](expressions/loop-expr.html#predicate-pattern-loops)
+* [`for` expressions](expressions/loop-expr.html#iterator-loops)
+* Inside other patterns
+
+## Destructuring
+
+Patterns can be used to *destructure* structs, enums, and tuples. Destructuring
+breaks a value up into its component pieces. The syntax used is almost the same as
+when creating such values. When destructing a data structure with named (but
+not numbered) fields, it is allowed to write `fieldname` as a shorthand for
+`fieldname: fieldname`. In a pattern whose head expression has a `struct`,
+`enum` or `tuple` type, a placeholder (`_`) stands for a *single* data field,
+whereas a wildcard `..` stands for *all* the remaining fields of a particular variant.
+
+```rust
+# enum Message {
+#     Quit,
+#     WriteString(String),
+#     Move { x: i32, y: i32 },
+#     ChangeColor(u8, u8, u8),
+# }
+# let message = Message::Quit;
+match message {
+    Message::Quit => println!("Quit"),
+    Message::WriteString(write) => println!("{}", &write),
+    Message::Move{ x, y: 0 } => println!("move {} horizontally", x),
+    Message::Move{ .. } => println!("other move"),
+    Message::ChangeColor { 0: red, 1: green, 2: _ } => {
+        println!("color change, red: {}, green: {}", red, green);
+    }
+};
+```
+
+## Refutability
+
+A pattern is said to be *Refutable* when it **has the possibility of not being matched**
```

Comment:
"Refutable" -> "refutable" (casing). I'm iffy about bolding the definition part. especially since the irrefutable definition part isn't bolded. If you want to propose changing our style to actually bold them like that, file an issue.